sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -0,0 +1,479 @@
+}
+
+static int rt711_sdca_interrupt_callback(struct sdw_slave *slave,
+					struct sdw_slave_intr_status *status)
+{
+	struct rt711_sdca_priv *rt711 = dev_get_drvdata(&slave->dev);
+	int ret = 0, stat, count = 0, retry = 10;
+	unsigned int sdca_cascade, scp_sdca_stat1, scp_sdca_stat2;
+	unsigned int buf;
+
+	dev_dbg(&slave->dev,
+		"%s control_port_stat=%x, sdca_cascade=%x", __func__,
+		status->control_port, status->sdca_cascade);
+
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &rt711->scp_sdca_stat1);
+	if (ret < 0)
+		goto io_error;
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &rt711->scp_sdca_stat2);
+	if (ret < 0)
+		goto io_error;
+
+	do {
+		/* clear flag */
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_0) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT1, SDW_SCP_SDCA_INTMASK_SDCA_0);
+			if (ret < 0)
+				goto io_error;
+		}
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_8) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT2, SDW_SCP_SDCA_INTMASK_SDCA_8);
+			if (ret < 0)
+				goto io_error;
+		}
+
+		/* check if flag clear or not */
+		ret = regmap_read(rt711->regmap, SDW_DP0_INT, &buf);
+		if (ret < 0)
+			goto io_error;
+		sdca_cascade = buf & SDW_DP0_SDCA_CASCADE;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat1 = buf & SDW_SCP_SDCA_INTMASK_SDCA_0;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat2 = buf & SDW_SCP_SDCA_INTMASK_SDCA_8;
+
+		stat = scp_sdca_stat1 || scp_sdca_stat2 || sdca_cascade;
+
+		count++;
+	} while (stat != 0 && count < retry);
+
+	if (count == retry) {
+		dev_warn(&slave->dev, "Reached MAX_RETRY on SDCA_Cascade clear\n");
+		dev_warn(&slave->dev,
+			"%s scp_sdca_stat1=0x%x, scp_sdca_stat2=0x%x\n", __func__,
+			rt711->scp_sdca_stat1, rt711->scp_sdca_stat2);
+	}
+
+	if (status->sdca_cascade)
+		mod_delayed_work(system_power_efficient_wq,
+			&rt711->jack_detect_work, msecs_to_jiffies(250));
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static struct sdw_slave_ops rt711_sdca_slave_ops = {
+	.read_prop = rt711_sdca_read_prop,
+	.interrupt_callback = rt711_sdca_interrupt_callback,
+	.update_status = rt711_sdca_update_status,
+};
+
+static int rt711_sdca_sdw_probe(struct sdw_slave *slave,
+				const struct sdw_device_id *id)
+{
+	struct regmap *sdw_regmap, *regmap;
+	int ret;
+
+	/* Regmap Initialization */
+	sdw_regmap = devm_regmap_init_sdw(slave, &rt711_sdca_sdw_regmap);
+	if (!sdw_regmap)
+		return -EINVAL;
+
+	regmap = devm_regmap_init(&slave->dev, NULL,
+		&slave->dev, &rt711_sdca_regmap);
+	if (IS_ERR(regmap))
+		return PTR_ERR(regmap);
+
+	ret = rt711_sdca_init(&slave->dev, sdw_regmap, regmap, slave);
```

Comment:
remove `ret` and just `return rt711_sdca_init(&slave->dev, sdw_regmap, regmap, slave);`